src/OrchardCore.Modules/OrchardCore.Roles/Migrations/SystemRolesMigrations.cs

```diff
@@ -0,0 +1,37 @@
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Logging;
+using OrchardCore.Data.Migration;
+using OrchardCore.Security;
+
+namespace OrchardCore.Roles.Migrations;
+
+public sealed class SystemRolesMigrations : DataMigration
+{
+    private readonly ISystemRoleNameProvider _systemRoleNameProvider;
+    private readonly RoleManager<IRole> _roleManager;
+    private readonly ILogger _logger;
+
+    public SystemRolesMigrations(
+        ISystemRoleNameProvider systemRoleNameProvider,
+        RoleManager<IRole> roleManager,
+        ILogger<RolesMigrations> logger)
+    {
+        _systemRoleNameProvider = systemRoleNameProvider;
+        _roleManager = roleManager;
+        _logger = logger;
+    }
+
+    public async Task<int> CreateAsync()
+    {
+        var systemRoles = await _systemRoleNameProvider.GetSystemRolesAsync();
```

Comment:
I am wondering if there is a raise condition somewhere.

Try implementing this logic using recipe step instead. You could do the same thing using IFeatureEventHandler

Reply:
Do you mean calling `RoleStep` inside the data migration? If yes the `RecipeExecutionContext` will be missing, we could find a way to pass it in but it will become more complicated

Reply:
I figure out there're roles in the test recipe :)